// @google-pay/button-angular/esm2020/button-angular/lib/google-pay-button.module.mjs
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { GooglePayButtonComponent } from './google-pay-button.component';
import { NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class GooglePayButtonModule {
}
GooglePayButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: GooglePayButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GooglePayButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.2", ngImport: i0, type: GooglePayButtonModule, declarations: [GooglePayButtonComponent], exports: [GooglePayButtonComponent] });
GooglePayButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: GooglePayButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: GooglePayButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [GooglePayButtonComponent],
                    imports: [],
                    exports: [GooglePayButtonComponent],
                }]
        }] });
